backend/search/models.py

```diff
@@ -40,54 +51,162 @@ class Meta:
-            except (OSError, UnicodeDecodeError) as err:
-                self.errors += 'Error searching database {}: ' \
-                    .format(database) + str(err) + '\n'
-                result = ''  # No break because completed_part is to be updated
-            try:
-                matches.extend(parse_search_result(
-                    result, self.component.slug, database)
+            if not stop_adding_results:
+                query = generate_xquery_search(
+                    database, self.xpath, self.variables
                 )
-            except ValueError as err:
-                self.errors += 'Error parsing search result in database {}: ' \
-                    '{}\n'.format(database, err)
-            self.number_of_results = len(matches)
-            self.results = json.dumps(matches)
+                try:
+                    result = basex.perform_query(query)
+                except (OSError, UnicodeDecodeError) as err:
+                    self.errors += 'Error searching database {}: ' \
+                        .format(database) + str(err) + '\n'
+                    result = ''  # No break, keep going
+                results_for_database = result.count('<match>')
+                maximum_to_add = settings.MAXIMUM_RESULTS_PER_COMPONENT - \
+                    self.number_of_results
+                self.number_of_results += results_for_database
+                if results_for_database > maximum_to_add:
+                    result = self._truncate_results(
+                        result, maximum_to_add
+                    )
+                if results_for_database >= maximum_to_add:
+                    stop_adding_results = True
+                resultsfile.write(result)
+            else:
+                # Only count from now on
+                query = generate_xquery_count(database, self.xpath)
+                try:
+                    count = int(basex.perform_query(query))
+                except (OSError, UnicodeDecodeError, ValueError) as err:
+                    self.errors += 'Error searching database {}: ' \
+                        .format(database) + str(err) + '\n'
+                    count = 0
+                self.number_of_results += count
             self.completed_part += size
             if timer() > next_save_time:
                 self.save()
                 next_save_time = timer() + 1
                 # Also check if search has been cancelled
                 if query_id is not None:
-                    cancelled = SearchQuery.objects \
+                    cancelled_value = SearchQuery.objects \
                         .values_list('cancelled', flat=True).get(id=query_id)
-                    if cancelled:
-                        return
-        self.search_completed = timezone.now()
+                    if cancelled_value:
+                        cancelled = True
+                        break
+        resultsfile.close()
```

Comment:
better to use a `with` statement, as uncaught exceptions within this function could leave the file open